src/backend/stream.rs

```diff
@@ -100,72 +98,70 @@ impl<'ctx> Stream<'ctx> {
+                s.peek()
             }
 
             logv!("Input callback buffer size {}", nbytes);
-            let mut stm = &mut *(u as *mut Stream);
+            let mut stm = unsafe { &mut *(u as *mut Stream) };
             if stm.shutdown {
                 return;
             }
 
-            let mut read_data: *const c_void = ptr::null();
-            let mut read_size: usize = 0;
-            while read_from_input(s, &mut read_data, &mut read_size) > 0 {
+            while let Ok((read_data, read_size)) = read_from_input(s) {
                 /* read_data can be NULL in case of a hole. */
                 if !read_data.is_null() {
-                    let in_frame_size = pa_frame_size(&stm.input_sample_spec);
+                    let in_frame_size = unsafe { pa_frame_size(&stm.input_sample_spec) };
                     let read_frames = read_size / in_frame_size;
 
                     if !stm.output_stream.is_null() {
                         // input/capture + output/playback operation
-                        let out_frame_size = pa_frame_size(&stm.output_sample_spec);
+                        let out_frame_size = unsafe { pa_frame_size(&stm.output_sample_spec) };
                         let write_size = read_frames * out_frame_size;
                         // Offer full duplex data for writing
-                        let stream = stm.output_stream;
+                        let stream = &stm.output_stream as *const _;
                         stm.trigger_user_callback(stream, read_data, write_size);
                     } else {
                         // input/capture only operation. Call callback directly
-                        let got = stm.data_callback.unwrap()(stm as *mut _ as *mut _,
-                                                             stm.user_ptr,
-                                                             read_data,
-                                                             ptr::null_mut(),
-                                                             read_frames as c_long);
+                        let got = unsafe {
+                            stm.data_callback.unwrap()(stm as *mut _ as *mut _,
+                                                       stm.user_ptr,
+                                                       read_data,
+                                                       ptr::null_mut(),
+                                                       read_frames as c_long)
+                        };
+
                         if got < 0 || got as usize != read_frames {
-                            pa_stream_cancel_write(s);
+                            let _ = s.cancel_write();
                             stm.shutdown = true;
                             break;
                         }
                     }
                 }
 
                 if read_size > 0 {
-                    pa_stream_drop(s);
+                    let _ = s.drop_record();
```

Comment:
In this case drop is rust destructor, so I can't have pa_stream.drop() mean pa_stream_drop().  Unless I remove implementing Drop for pulse::Stream, but drop() still means something in Rust.